feat(zellij-utils): Add source `client_id` to pipe messages

EmrysMyrddin opened the pull request:
# Proposal
## Description
Adds a the `client_id` of the client being the source of the pipe message.

## Problem

I want to create a plugin that allows to write chars to a specific pane based on it's name.
For this, I want my plugin to react to a pipe message (which can be sent from the CLI or a KeyBind), then tries to find a pane with a matching title in the focused tab.

But, since a pipe message is always sent to all clients, even when it is triggered by a KeyBind, I can't know which tab is focused by the client how sent the KeyBind.

# Solution

After a lot of trial and error, I think that adding the source `client_id` to the `PipeMessage` structure is the minimal way to fix my problem.

With the source `client_id`, my plugin can know from which tab the pipe message was sent by keeping track of which tab each client is currently focusing. 

If it's coming from the CLI, my plugin can look for a `pane_id` argument in the pipe message. The CLI caller can send the current pane id from `ZELLIJ_PANE_ID` env variable.

It's not perfect (the cli source is a bit cumbersome), but it's minimal, and allows for way better behavior in multiplayer mode.

imsnif commented:
Hey, thanks for this and good find. I think it makes sense all in all, but I'd rather not add the cli_client_id, since it's misleading (these clients are usually extremely short lived and don't have many of the other guarantees clients have). We could do this by making the field optional in the protobuf declaration.

Also, if you'd like you can add the `PaneId` (also as an optional field) to the message coming from the CLI to perform the tracking you suggested.

EmrysMyrddin commented:
@imsnif I've updated the PR to only send the `client_id` when it's not a CLI client.
I've also added the `pane_id` as you suggested, so that Plugin devs don't have to manually track the client focused pane just to know which pane was focused when a KeyBind was pressed.

Since I'm already modifying the protobuf, perhaps we can merge #4095 into this PR ? Since it's already a breaking change, we can take the opportunity to clean up this.

imsnif commented:
Hey, from a cursory look this seems fine - thanks! I appreciate the thought with the protobuffs in xtask, but I'd rather not conflate things. Can we remove it for now?

As for the `PaneId` - if we just add stuff to protobuf files, it's not a breaking change. It's only a breaking change if we change stuff. As I mentioned in the previous PR, changing stuff around might or might not be a breaking change (might also be a change in some implementations and not in others) - but I'd really rather not spend time on this to find out. My resources are limited and I try to spend very deliberate time on changes/additions that will benefit users/plugin developers. Thanks for understanding!

EmrysMyrddin commented:
Ok I see :-) Actually It seems I can use the `PaneId` defined in `event.proto`, so it's not mandatory to move things.

For the xtask thing, sure I will remove it. Do you want me to open a new PR for it ?

After testing, it seems the PaneId is not set for `KeybindPipe` action. Do you know why ? Should I change this too, or you prefere the `pane_id` to be set only for the `CliPipe` action ?

EmrysMyrddin commented:
@imsnif May I ask you if you have time to take a new look at this PR ?

imsnif commented:
My apologies, I am juggling a lot of things and am merely Human. Best bet is to try to make this change as small and concise as possible so that I can review and merge it in one go. I'll try to take a look in the next few weeks. Thanks for understanding.

EmrysMyrddin commented:
No problem, I fully understand that :-) Maintaining alone this kind of project is not an easy task!

I've cleaned up the PR. The changes should be as minimal as I can think about.

imsnif commented on zellij-server/src/plugins/mod.rs:
```diff
@@ -719,9 +724,12 @@ pub(crate) fn plugin_thread_main(
                         );
                     },
                     None => {
+                        log::warn!("PIPE MESSAGE TO ALL PLUGINS");
```
Let's please remove this

imsnif commented on zellij-server/src/plugins/mod.rs:
```diff
@@ -754,14 +763,24 @@ pub(crate) fn plugin_thread_main(
                             // send to specific plugin(s)
                             pipe_to_specific_plugins(
                                 PipeSource::Keybind,
+                                Some(cli_client_id),
```
I think this should be a `None`. Even though in this case cli_client_id is will always be None (since this is coming from a keybinding), we shouldn't re-use it in case this behavior somehow changes in the future. Then, if we want to add the cli_client_id we'll do it explicitly.

imsnif commented on zellij-server/src/plugins/mod.rs:
```diff
@@ -785,6 +804,8 @@ pub(crate) fn plugin_thread_main(
                             // no specific destination, send to all plugins
                             pipe_to_all_plugins(
                                 PipeSource::Keybind,
+                                Some(cli_client_id),
```
Same here

imsnif commented on zellij-server/src/plugins/unit/plugin_tests.rs:
```diff
@@ -6191,6 +6191,7 @@ pub fn unblock_input_plugin_command() {
     std::thread::sleep(std::time::Duration::from_millis(500));
 
     let _ = plugin_thread_sender.send(PluginInstruction::CliPipe {
+        pane_id: None,
```
Let's maybe make one of these a `Some` for the sake of completeness in the tests?

imsnif commented on zellij-utils/src/plugin_api/pipe_message.proto:
```diff
@@ -10,6 +12,8 @@ message PipeMessage {
     repeated Arg args = 6;
     bool is_private = 7;
+    optional uint32 clientId = 8;
+    optional api.event.PaneId paneId = 9;
```
These should be snake case

imsnif reviewed (changes requested):
Thank you for your patience! I'll get the CI to run the tests and I also left some comments.

EmrysMyrddin commented:
By the way I don't know if you want me to clean up the commit history or if you usually squash on merge?

EmrysMyrddin commented:
Hi!

>thank you for your endless patience!
Ahah no problem 🙂 You are already giving a lot to the community by making zellij!

Yes, I would still be happy to work on this :-) I still have the project to use this for Helix interoparibility with other tools (waiting for them to have a proper plugin system, which will probably not happen any time soon)

> In particular we're not tracking client_ids for cli_pipes natively, eg. here: [main/zellij-server/src/route.rs#L1744](https://github.com/zellij-org/zellij/blob/main/zellij-server/src/route.rs?rgh-link-date=2026-01-14T07%3A48%3A12Z#L1744)

I'm not sure to understand what you mean?

EmrysMyrddin commented:
Hi @imsnif! Sorry, I didn't have much time to work on this lately.

I've made a rebase, it should now be mergeable :-)

As for the `cli_client_id`, you made the following comment: https://github.com/zellij-org/zellij/pull/4094#discussion_r2375838735

From what I understand of the code, as of today, I should revert the change that removed the `Some(cli_client_id)` as the `source_client_id` in the case of a `KeyBind` event. Because now, this `cli_client_id` should pretty much always makes sense ?

imsnif commented:
Fair question. From my point of view: I am unable to tell the future or what we'll be using this field for if we decide to use it. I'd rather the person implementing that change will make an explicit decision than have an implicit one unknowingly made for them.
